### ISSUE TITLE
mac_os_x brew error: use of undeclared identifier 'iterator'

### ISSUE THREAD

**Issue by OnlyOneCannolo** [NASA-SW-VnV/ikos#79]
Steps to reproduce:
```bash
$ brew update
$ brew upgrade
$ brew install nasa-sw-vnv/core/ikos
```

Error message:
```bash
==> make install
Last 15 lines from /Users/*/Library/Logs/Homebrew/ikos/05.make:
                                                                ^
/usr/local/opt/llvm/include/llvm/IR/DebugInfoMetadata.h:147:61: error: use of undeclared identifier 'iterator'; did you mean 'operator'?
  iterator end() const { return N ? iterator(N->op_end()) : iterator(); }
                                                            ^
/usr/local/opt/llvm/include/llvm/IR/DebugInfoMetadata.h:160:39: error: unknown type name 'StorageType'
  DINode(LLVMContext &C, unsigned ID, StorageType Storage, unsigned Tag,
                                      ^
/usr/local/opt/llvm/include/llvm/IR/DebugInfoMetadata.h:161:19: error: use of undeclared identifier 'Metadata'
         ArrayRef<Metadata *> Ops1, ArrayRef<Metadata *> Ops2 = None)
                  ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [analyzer/CMakeFiles/ikos-analyzer.dir/src/ikos_analyzer.cpp.o] Error 1
make[1]: *** [analyzer/CMakeFiles/ikos-analyzer.dir/all] Error 2
make: *** [all] Error 2
```

System info:
```bash
$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.14.2
BuildVersion:	18C54
```

**Comment by arthaud** [comment]
Hi @OnlyOneCannolo,

Unfortunately, I'm not able to reproduce the bug.

I tried on my work laptop (macOS 10.13.6) using a fresh Homebrew, and everything works as expected.
I tried on Travis CI (macOS 10.13.3) and it works too:
https://travis-ci.org/NASA-SW-VnV/homebrew-core

Could you try `brew install nasa-sw-vnv/core/ikos --debug --verbose`?
Could you also check the output of `brew doctor`?

**Comment by OnlyOneCannolo** [comment]
I created a [gist](https://gist.github.com/OnlyOneCannolo/fd6629ffa61c93cc963ab867d4fd60af) with the results of running those commands. The results of `brew doctor` appear to be unrelated to any IKOS dependencies.

**Comment by arthaud** [comment]
Thank you.

I see the following error in the output:
```
[ 41%] Linking CXX executable ikos-import
/tmp/ikos-20190112-61885-14p58jz/ikos-2.1/analyzer/src/analysis/call_context.cpp:60:16: error: no member named 'try_emplace' in 'llvm::DenseMap<std::__1::pair<ikos::analyzer::CallContext *, ikos::ar::CallBase *>, std::__1::unique_ptr<ikos::analyzer::CallContext, std::__1::default_delete<ikos::analyzer::CallContext> >, llvm::DenseMapInfo<std::__1::pair<ikos::analyzer::CallContext *, ikos::ar::CallBase *> > >'
    this->_map.try_emplace({parent, call},
    ~~~~~~~~~~ ^
1 error generated.
```

The problem might be that you have an older version of llvm installed at `/usr/local/opt/llvm`.

Could you try to reinstall llvm 7?

Also, check if `/usr/local/opt/llvm` points to `/usr/local/Cellar/llvm/7.0.1`:
```
$ ls -l /usr/local/opt/llvm
/usr/local/opt/llvm -> ../Cellar/llvm/7.0.1
```

**Comment by OnlyOneCannolo** [comment]
```bash
$ brew reinstall llvm@7
$ ls -l /usr/local/opt/llvm
/usr/local/opt/llvm -> ../Cellar/llvm/7.0.1
$ echo $LDFLAGS
-L/usr/local/opt/llvm/lib
$ echo $PATH
/usr/local/opt/llvm/bin:#...
```
The results of `brew doctor` and `brew install nasa-sw-vnv/core/ikos --debug --verbose` are the same as last time.

I may try building from source instead. Thank you for looking into this.

**Comment by arthaud** [comment]
I just released IKOS 2.2.
I updated the Homebrew Formula. Hopefully, it fixes the problem.
I'm closing this issue. Please create a new issue if you have the same problem again.
